### PR TITLE
SI-6111 accept single-subpattern unapply pattern

### PR DESCRIPTION
**Pull request by adriaanm**
An extractor pattern `X(p)` should type check for any `X.unapply`/`X.unapplySeq`
that returns an `Option[_]` -- previously we were confused about the case where
it was an `Option[(T1, ... , Tn)]`. In this case, the expected type for the
pattern `p` is simply `(T1, ... , Tn)`.

While I was at it, tried to clean up unapplyTypeList and friends (by replacing them by extractorFormalTypes).

From the spec:
   8.1.8 ExtractorPatterns

   An extractor pattern x(p1, ..., pn) where n ≥ 0 is of the same syntactic form as a constructor pattern.
   However, instead of a case class, the stable identifier x denotes an object which has a member method named unapply or unapplySeq that matches the pattern.
   An unapply method in an object x matches the pattern x(p1, ..., pn) if it takes exactly one argument and one of the following applies:

   n = 0 and unapply’s result type is Boolean.

   n = 1 and unapply’s result type is Option[T], for some type T.
     the (only) argument pattern p1 is typed in turn with expected type T

   n > 1 and unapply’s result type is Option[(T1, ..., Tn)], for some types T1, ..., Tn.
     the argument patterns p1, ..., pn are typed in turn with expected types T1, ..., Tn

### CODE REVIEW

**Comment by heathermiller** [comment]
Amazing! Thank you Adriaan!!

**Comment by adriaanm** [comment]
review by @paulp prettyPlease

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -2749,6 +2749,7 @@ trait Typers extends Modes with Adaptations with Tags {
     def typedArgs(args: List[Tree], mode: Int) =
       args mapConserve (arg => typedArg(arg, mode, 0, WildcardType))
 
+    // [adriaan] as far as I can tell, formals0 is only supplied to detect whether the last formal was originally a vararg
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: It also uses formals to take BYVALmode off when typing by-name arguments .

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 2]
```diff
@@ -49,6 +49,70 @@ trait Infer {
...
+
+    val formals =
+      if (nbSubPats == 0 && booleanExtractor && !isUnapplySeq)  Nil
+      else resTp.baseType(OptionClass).typeArgs match {
+        case optionTArg :: Nil =>
+          if (nbSubPats == 1)
+            if (isUnapplySeq) List(seqToRepeatedChecked(optionTArg))
+            else List(optionTArg)
+          // in principle, the spec doesn't allow just any subtype of Product, it *must* be TupleN[...] -- see run/virtpatmat_extends_product.scala
+          // should check `isTupleType(optionTArg)` -- this breaks plenty of stuff, though...
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: Didn't we agree long ago that it should all be specified in terms of Product, not Tuple?

**Comment by paulp** [comment]
This looks okay to me.  Is there some good reason we even need Tuple1 as a separate entity from Some? That is, if Some extended Product1, which I observe right now it does not.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 1 above)
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: right, I'll write some proper docs and submit a new commit

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 2 above)
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: we probably don't have a choice at this point, so I'll add a TODO to update the spec

**Pull request merged**